pkg/api/validate/openshiftcluster_validate.go

```diff
@@ -0,0 +1,133 @@
+package validate
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/api/validate/dynamic"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/aad"
+	"github.com/Azure/ARO-RP/pkg/util/refreshable"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+	"github.com/Azure/ARO-RP/pkg/util/subnet"
+)
+
+// OpenShiftClusterValidator is the dynamic validator interface
+type OpenShiftClusterValidator interface {
+	Dynamic(context.Context) error
+}
+
+// NewOpenShiftClusterValidator creates a new OpenShiftClusterValidator
+func NewOpenShiftClusterValidator(log *logrus.Entry, env env.Core, oc *api.OpenShiftCluster, subscriptionDoc *api.SubscriptionDocument, fpAuthorizer refreshable.Authorizer) OpenShiftClusterValidator {
+	return &openShiftClusterValidator{
+		log: log,
+		env: env,
+
+		oc:              oc,
+		subscriptionDoc: subscriptionDoc,
+		fpAuthorizer:    fpAuthorizer,
+	}
+}
+
+type openShiftClusterValidator struct {
+	log *logrus.Entry
+	env env.Core
+
+	oc              *api.OpenShiftCluster
+	subscriptionDoc *api.SubscriptionDocument
+	fpAuthorizer    refreshable.Authorizer
+}
+
+// Dynamic validates an OpenShift cluster
+func (dv *openShiftClusterValidator) Dynamic(ctx context.Context) error {
+	// Get all subnets
+	subnetIDs := []string{dv.oc.Properties.MasterProfile.SubnetID}
+
+	for _, s := range dv.oc.Properties.WorkerProfiles {
+		subnetIDs = append(subnetIDs, s.SubnetID)
+	}
+
+	// its callers responsibility to provive unique values for validations
+	// if this is required. In this case on create worker pool profiles are
+	// not enriched. During cluster runtime they get enriched and contains multiple
+	// duplicate values for multiple worker pools
```

Comment:
not sure I agree.  For most validators, uniqing is an optimisation which they choose to do, and should do as an implementation detail.  For the CIDR validator, uniqing is a requirement - let the validator do that.